src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot.aarch64/src/jdk/vm/ci/hotspot/aarch64/AArch64HotSpotRegisterConfig.java

```diff
@@ -264,12 +267,29 @@ private CallingConvention callingConvention(RegisterArray generalParameterRegist
             }
 
             if (locations[i] == null) {
-                ValueKind<?> valueKind = valueKindFactory.getValueKind(kind);
-                locations[i] = StackSlot.get(valueKind, currentStackOffset, !type.out);
-                currentStackOffset += Math.max(valueKind.getPlatformKind().getSizeInBytes(), target.wordSize);
+                if (type == HotSpotCallingConventionType.NativeCall) {
+                    ValueKind<?> valueKind = valueKindFactory.getValueKind(kind);
+                    int kindSize = valueKind.getPlatformKind().getSizeInBytes();
+                    if (macOS && currentStackOffset % kindSize != 0) {
+                        // In MacOS natural alignment is used
+                        // See https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms
+                        currentStackOffset += kindSize - currentStackOffset % kindSize;
+                    }
+                    locations[i] = StackSlot.get(valueKind, currentStackOffset, !type.out);
+                    if (macOS) {
+                        // In MacOS "Function arguments may consume slots on the stack that are not multiples of 8 bytes"
+                        // See https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms
+                        currentStackOffset += kindSize;
+                    } else {
+                        currentStackOffset += Math.max(kindSize, target.wordSize);
+                    }
+                } else {
+                    ValueKind<?> valueKind = valueKindFactory.getValueKind(kind);
+                    locations[i] = StackSlot.get(valueKind, currentStackOffset, !type.out);
+                    currentStackOffset += Math.max(valueKind.getPlatformKind().getSizeInBytes(), target.wordSize);
+                }
             }
```

Comment:
So I'm curious: why not subclass `AArch64HotSpotRegisterConfig` here, or maybe even use an interface, rather than the boolean?

Reply:
I tried to be closer to the original review https://github.com/openjdk/jdk/pull/6641 that requires only 2 fixes and tried to do only this in order to continue easily.

Could you clarify please what boolean you talk about? `private final boolean macOS;` that was pushed into `class AArch64HotSpotRegisterConfig`, right? I'm hesitating a bit because of the highlighted code.

Reply:
Yes, that `macOS` boolean.

Maybe it's not worth the effort, but it seems to me as though the use of the boolean in several places is something of a code smell, and this patch makes it more so. The control flow is not easy to follow
I am wondering if refactoring it so that the code between L269 and L291 were broken out into two methods, one fot MacOS and one for the others. I  might be wrong, but I'd try it.

Reply:
Thanks for the tip! Absolutely agree, it's worth doing refactoring here. I'll try to follow this way.

Reply:
@theRealAph could you please take a look! I've made some additional refactoring in order to get rid of those "boolean"s where it's possible. I moved them from AArch64HotSpotVMConfig to TargetDescription. This makes an access to them easier across the code. As a 2nd step it's possible to substitute these "boolean"s with "enum" if it's needed. But imho it's better to keep them as "boolean"s at the moment.
In addition I see a new class RISCV64HotSpotVMConfig (committed after this pr) that also declared the same boolean:
`final boolean linuxOs = Services.getSavedProperty("os.name", "").startsWith("Linux");`
If this refactoring is made, this "boolean" won't be needed as well because it can be accessed from TargetDescription.

Reply:
This looks reasonable enough.